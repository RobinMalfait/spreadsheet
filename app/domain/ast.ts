import type { Location } from '~/domain/expression'

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  raw: string
  loc: Location
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  raw: string
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}
