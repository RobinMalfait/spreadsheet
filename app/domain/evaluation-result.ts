export enum EvaluationResultKind {
  ERROR = 'ERROR',
  EMPTY = 'EMPTY',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.ERROR; value: string }
  | { kind: EvaluationResultKind.EMPTY; value: string }
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }
