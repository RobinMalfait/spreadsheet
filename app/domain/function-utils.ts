import type { EvaluationResult } from '~/domain/evaluation-result'
import { parse } from '~/domain/signature/parser'
import { tokenize } from '~/domain/signature/tokenizer'
import { validate } from '~/domain/signature/validate'

export function expose<
  T extends EvaluationResult,
  R extends EvaluationResult | EvaluationResult[],
>(signature: string, handle: (...args: T[]) => R): (...args: T[]) => R {
  let sig = parse(tokenize(signature))

  return Object.assign(
    (...args: T[]) => {
      let errors = validate(sig, args)
      if (errors) return errors as R

      // Handle
      return handle(...args) as R
    },
    { signature: sig },
  )
}
