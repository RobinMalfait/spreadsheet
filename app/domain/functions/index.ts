export * from './binary'
export * from './date'
export * from './logic'
export * from './lookup'
export * from './math'
export * from './privileged'
export * from './sequence'
export * from './statistics'
export * from './text'
export * from './types'
