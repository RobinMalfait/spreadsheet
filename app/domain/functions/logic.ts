import { type EvaluationResult, EvaluationResultKind } from '~/domain/evaluation-result'
import { expose } from '~/domain/function-utils'

export const TRUE = expose(
  `
    @description The boolean value true

    TRUE()
  `,
  () => {
    return { kind: EvaluationResultKind.BOOLEAN, value: true }
  },
)

export const FALSE = expose(
  `
    @description The boolean value false

    FALSE()
  `,
  () => {
    return { kind: EvaluationResultKind.BOOLEAN, value: false }
  },
)

// @describe test The condition to evaluate
// @describe consequent The value to return if the condition is true
// @describe alternate The value to return if the condition is false
export const IF = expose(
  `
    @description Returns one value if a condition is true and another value if it is false

    IF(test: BOOLEAN, consequent: T, alternate: T)
  `,
  (test: EvaluationResult, consequent: EvaluationResult, alternate: EvaluationResult) => {
    return test.value ? consequent : alternate
  },
)

// @describe expressions The conditions to evaluate
export const AND = expose(
  `
    @description Returns true if all conditions are true

    AND(...expressions: T)
  `,
  (...args) => {
    return args.every((arg) => arg.value) ? TRUE() : FALSE()
  },
)

// @describe expressions The conditions to evaluate
export const OR = expose(
  `
    @description Returns true if any condition is true

    OR(...expressions: T)
  `,
  (...args) => {
    return args.some((arg) => arg.value) ? TRUE() : FALSE()
  },
)

// @describe value The condition to negate
export const NOT = expose(
  `
    @description Returns true if the condition is false

    NOT(value: BOOLEAN)
  `,
  (lhs: EvaluationResult) => {
    return lhs.value ? FALSE() : TRUE()
  },
)
