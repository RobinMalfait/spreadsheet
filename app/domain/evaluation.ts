import { format } from 'date-fns'
import { type AST, AstKind, BinaryExpressionOperator } from '~/domain/ast'
import * as functions from '~/domain/functions'
import type { Spreadsheet } from '~/domain/spreadsheet'
import { expandRange } from '~/domain/walk-ast'

export enum EvaluationResultKind {
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }

export function evaluateExpression(
  ast: AST,
  spreadsheet: Spreadsheet,
): EvaluationResult[] {
  switch (ast.kind) {
    case AstKind.NUMBER_LITERAL:
      return [{ kind: EvaluationResultKind.NUMBER, value: ast.value }]

    case AstKind.STRING_LITERAL: {
      if (ast.value.trim() === '') {
        return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
      }

      // Try to coerce the string to a number, and if it works, return a number
      // instead of a string
      let asNumber = Number(ast.value)
      if (!Number.isNaN(asNumber)) {
        return [{ kind: EvaluationResultKind.NUMBER, value: asNumber }]
      }

      return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
    }

    case AstKind.CELL:
      return spreadsheet.evaluate(ast.name)

    case AstKind.RANGE: {
      let out = []
      for (let cell of expandRange(ast)) {
        for (let child of spreadsheet.evaluate(cell)) {
          out.push(child)
        }
      }
      return out
    }

    case AstKind.BINARY_EXPRESSION: {
      let lhs = evaluateExpression(ast.lhs, spreadsheet)
      let rhs = evaluateExpression(ast.rhs, spreadsheet)

      if (lhs.length !== 1 || rhs.length !== 1) {
        throw new Error('Expected a single result from each side')
      }

      let left = lhs[0]
      let right = rhs[0]

      if (
        left?.kind === EvaluationResultKind.NUMBER &&
        right?.kind === EvaluationResultKind.NUMBER
      ) {
        switch (ast.operator) {
          // Math operators
          case BinaryExpressionOperator.EXPONENT:
            return [functions.POWER(left, right)]
          case BinaryExpressionOperator.MULTIPLY:
            return [functions.MULTIPLY(left, right)]
          case BinaryExpressionOperator.DIVIDE:
            return [functions.DIVIDE(left, right)]
          case BinaryExpressionOperator.ADD:
            return [functions.SUM(left, right)]
          case BinaryExpressionOperator.SUBTRACT:
            return [functions.SUBTRACT(left, right)]

          // Comparison operators
          case BinaryExpressionOperator.EQUALS:
            return left.value === right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.NOT_EQUALS:
            return left.value !== right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN:
            return left.value < right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN_EQUALS:
            return left.value <= right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN:
            return left.value > right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN_EQUALS:
            return left.value >= right.value ? [functions.TRUE()] : [functions.FALSE()]
          default:
            ast.operator satisfies never
        }
      }

      throw new Error(`Invalid operation \`${ast.kind}\``)
    }

    case AstKind.FUNCTION: {
      if (!Object.hasOwn(functions, ast.name)) {
        throw new Error(`Unknown function \`${ast.name}\``)
      }

      let fn = functions[ast.name as keyof typeof functions]
      let args = ast.args.flatMap((arg) => evaluateExpression(arg, spreadsheet))
      // @ts-expect-error Some functions have a different arity, but we're not
      // checking that here.
      let result = fn(...args)
      if (Array.isArray(result)) {
        return result
      }
      return [result]
    }
  }
}

export function printEvaluationResult(result: EvaluationResult): string {
  switch (result.kind) {
    case EvaluationResultKind.NUMBER:
      return result.value.toString()
    case EvaluationResultKind.STRING:
      return result.value
    case EvaluationResultKind.BOOLEAN:
      return result.value ? 'TRUE' : 'FALSE'
    case EvaluationResultKind.DATETIME:
      if (result.date && result.time) return format(result.value, 'yyyy-MM-dd HH:mm:ss')
      if (result.date) return format(result.value, 'yyyy-MM-dd')
      return format(result.value, 'HH:mm:ss')
    default:
      result satisfies never
      return ''
  }
}
