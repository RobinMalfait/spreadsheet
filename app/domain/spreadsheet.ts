import { type AST, AstKind } from '~/domain/ast'
import { type EvaluationResult, evaluateExpression } from '~/domain/evaluation'
import { parse } from '~/domain/expression'
import * as functions from '~/domain/functions'
import { tokenize } from '~/domain/tokenizer'
import { WalkAction, walk } from '~/domain/walk-ast'
import { DefaultMap } from '~/utils/default-map'

export enum ComputationResultKind {
  VALUE = 'VALUE',
  ERROR = 'ERROR',
}

type ComputationValue = {
  kind: ComputationResultKind.VALUE
  value: EvaluationResult
}

type ComputationError = {
  kind: ComputationResultKind.ERROR
  short: string
  message: string
}

export class Spreadsheet {
  // Track each individual cell and it's contents. AST is pre-parsed.
  private cells = new Map<string, [raw: string, ast: AST]>()

  // Track all dependencies for each cell.
  private _dependencies = new DefaultMap<string, Set<string>>(() => new Set<string>())

  private evaluationCache = new Map<string, EvaluationResult[]>()

  get cellNames() {
    return Array.from(this.cells.keys())
  }

  has(cell: string): boolean {
    return this.cells.has(cell)
  }

  get(cell: string): string {
    let result = this.cells.get(cell)
    if (result) {
      return result[0]
    }
    return ''
  }

  set(cell: string, value: string) {
    // Reset state
    this.cells.delete(cell)

    // Clear existing dependencies for this cell
    let dependencies = this._dependencies.get(cell)
    dependencies.clear()

    // Clear the full evaluation cache. Very naive, but let's re-compute
    // everything the moment _something_ changes.
    this.evaluationCache.clear()

    // Value must be filled in
    if (value.trim() === '') return

    // Expression should exist
    let expression = value[0] === '=' ? value.slice(1) : `"${value}"`
    if (expression.trim() === '') return

    let tokens = tokenize(expression)
    let ast = parse(tokens)

    // Track all references in the AST
    walk([ast], (node) => {
      if (node.kind === AstKind.CELL) {
        dependencies.add(node.name)
      }

      return WalkAction.Continue
    })

    this.cells.set(cell, [value, ast])
  }

  functions() {
    let collator = new Intl.Collator('en', { sensitivity: 'base' })
    return Object.keys(functions).sort(collator.compare)
  }

  dependencies(cell: string): Set<string> {
    return new Set(this._dependencies.get(cell))
  }

  compute(cell: string): ComputationValue | ComputationError | null {
    try {
      let evaluationResult = this.evaluate(cell)
      if (evaluationResult.length > 1) {
        throw Object.assign(
          new Error(`Expected a single result, got ${evaluationResult.length}`),
          { short: '#VALUE!' },
        )
      }

      if (evaluationResult.length === 1) {
        return {
          kind: ComputationResultKind.VALUE,
          // biome-ignore lint/style/noNonNullAssertion: we already verified that the evaluationResult has at least one element.
          value: evaluationResult[0]!,
        }
      }

      return null
    } catch (err: unknown) {
      return {
        kind: ComputationResultKind.ERROR,
        // @ts-expect-error This is fineâ€¦
        short: err?.short ?? 'Error',
        message: (err as Error).message,
      }
    }
  }

  evaluate(cell: string): EvaluationResult[] {
    let result = this.cells.get(cell)
    if (!result) return []

    let cached = this.evaluationCache.get(cell)
    if (cached) return cached

    // TODO: Should this be moved to the `set` method?
    if (result[1].kind === AstKind.RANGE) {
      throw Object.assign(new Error('Cannot reference a range to a cell'), {
        short: '#VALUE!',
      })
    }

    // Verify references
    let dependencies = this._dependencies.get(cell)
    if (dependencies.size > 0) {
      let handled = new Set<string>()
      let todo = Array.from(dependencies)

      // Track where we came from
      let previous = cell

      while (todo.length > 0) {
        // biome-ignore lint/style/noNonNullAssertion: We verified that there is at least a single element in the while condition above.
        let next = todo.shift()!

        // No need to re-verify cells we've already handled
        if (handled.has(next)) continue
        handled.add(next)

        if (next === cell) {
          throw Object.assign(
            new Error(`Circular reference detected in cell ${previous}`),
            { short: '#REF!' },
          )
        }

        if (this.cells.has(next)) {
          // Track where we came from
          previous = next

          // Check transitive dependencies
          for (let other of this._dependencies.get(next)) {
            if (!handled.has(other)) {
              todo.push(other)
            }
          }
        }
      }
    }

    let out = evaluateExpression(result[1], this)
    this.evaluationCache.set(cell, out)
    return out
  }
}
