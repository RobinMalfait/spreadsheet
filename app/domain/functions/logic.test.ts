import { describe, expect, it } from 'vitest'
import { Spreadsheet } from '~/domain/spreadsheet'
import { visualizeSpreadsheet } from '~/test/utils'
import { exampleTests } from '~/test/utils'
import * as functions from './logic'
await exampleTests(functions)

describe('TRUE()', () => {
  it('should error when providing an argument', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TRUE(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: TRUE() does not take any arguments
      "
    `)
  })

  it('should result in TRUE', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TRUE()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬──────┐
      │   │ A    │
      ├───┼──────┤
      │ 1 │ TRUE │
      └───┴──────┘
      "
    `)
  })
})

describe('FALSE()', () => {
  it('should error when providing an argument', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=FALSE(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: FALSE() does not take any arguments
      "
    `)
  })

  it('should result in FALSE', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=FALSE()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ FALSE │
      └───┴───────┘
      "
    `)
  })
})

describe('IF()', () => {
  it('should error when the test is a literal string', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF("UH OH", 1, 2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: IF(test: BOOLEAN, consequent: T, alternate: T) Argument \`test\` received a \`STRING\`
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF(TRUE(), 1, 2, 3)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: IF(test: BOOLEAN, consequent: T, alternate: T) takes at most 3 arguments, got 4
      "
    `)
  })

  it('should result in the consequent when the test is truthy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=5')
    spreadsheet.set('B1', '=IF(A1 == 5, "Five!", "Not five!")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬─────────┐
      │   │ A │ B       │
      ├───┼───┼─────────┤
      │ 1 │ 5 │ "Five!" │
      └───┴───┴─────────┘
      "
    `)
  })

  it('should result in the alternate when the test is falsy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=IF(A1 == 5, "Five!", "Not five!")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬─────────────┐
      │   │ A │ B           │
      ├───┼───┼─────────────┤
      │ 1 │ 2 │ "Not five!" │
      └───┴───┴─────────────┘
      "
    `)
  })
})

describe('IF()', () => {
  it('should error when no value is provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SWITCH()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SWITCH(value: T, ...cases: T, default?: T) Argument \`value\` was not provided
      "
    `)
  })

  it('should error when no cases are provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SWITCH(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SWITCH(value: T, ...cases: T, default?: T) Argument \`cases\` was not provided
      "
    `)
  })

  it('should error when no case matches and no default was provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SWITCH("ABC", "FOO", 1, "BAR", 2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SWITCH(ABC) No matching case found
      "
    `)
  })

  it('should find the matching case, and return its value', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SWITCH(2, 1, "One", 2, "Two", 3, "Three")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ "Two" │
      └───┴───────┘
      "
    `)
  })

  it('should result in the default value when no match is found', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SWITCH(8, 1, "One", 2, "Two", 3, "Three", "Other")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────┐
      │   │ A       │
      ├───┼─────────┤
      │ 1 │ "Other" │
      └───┴─────────┘
      "
    `)
  })
})

describe('IF_ERROR()', () => {
  it('should error when the value is not provided at all', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF_ERROR()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: IF_ERROR(value: T, fallback: T) Argument \`value\` was not provided
      "
    `)
  })

  it('should error when the fallback is not provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF_ERROR(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: IF_ERROR(value: T, fallback: T) Argument \`fallback\` was not provided
      "
    `)
  })

  it('should result in the value itself if it is not an error', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF_ERROR("Not an error", "Fallback")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────────────────┐
      │   │ A              │
      ├───┼────────────────┤
      │ 1 │ "Not an error" │
      └───┴────────────────┘
      "
    `)
  })

  it('should result in the fallback if the value is an error', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=IF_ERROR(123 / 0, "Fallback")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────────────┐
      │   │ A          │
      ├───┼────────────┤
      │ 1 │ "Fallback" │
      └───┴────────────┘
      "
    `)
  })
})

describe('AND()', () => {
  it('should result in TRUE when all arguments are truthy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=1')
    spreadsheet.set('B1', '=2')
    spreadsheet.set('C1', '=3')
    spreadsheet.set('A2', '=AND(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬──────┬───┬───┐
      │   │ A    │ B │ C │
      ├───┼──────┼───┼───┤
      │ 1 │ 1    │ 2 │ 3 │
      ├───┼──────┼───┼───┤
      │ 2 │ TRUE │   │   │
      └───┴──────┴───┴───┘
      "
    `)
  })

  it('should result in FALSE when not all arguments are truthy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=1')
    spreadsheet.set('B1', '=0')
    spreadsheet.set('C1', '=3')
    spreadsheet.set('A2', '=AND(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┬───┬───┐
      │   │ A     │ B │ C │
      ├───┼───────┼───┼───┤
      │ 1 │ 1     │ 0 │ 3 │
      ├───┼───────┼───┼───┤
      │ 2 │ FALSE │   │   │
      └───┴───────┴───┴───┘
      "
    `)
  })
})

describe('OR()', () => {
  it('should result in TRUE when some arguments are truthy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=1')
    spreadsheet.set('B1', '=0')
    spreadsheet.set('C1', '=0')
    spreadsheet.set('A2', '=OR(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬──────┬───┬───┐
      │   │ A    │ B │ C │
      ├───┼──────┼───┼───┤
      │ 1 │ 1    │ 0 │ 0 │
      ├───┼──────┼───┼───┤
      │ 2 │ TRUE │   │   │
      └───┴──────┴───┴───┘
      "
    `)
  })

  it('should result in FALSE when no arguments are truthy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=0')
    spreadsheet.set('B1', '=0')
    spreadsheet.set('C1', '=0')
    spreadsheet.set('A2', '=OR(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┬───┬───┐
      │   │ A     │ B │ C │
      ├───┼───────┼───┼───┤
      │ 1 │ 0     │ 0 │ 0 │
      ├───┼───────┼───┼───┤
      │ 2 │ FALSE │   │   │
      └───┴───────┴───┴───┘
      "
    `)
  })
})

describe('NOT()', () => {
  it('should error when the test is a literal string', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOT("UH OH")')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: NOT(value: BOOLEAN) Argument \`value\` received a \`STRING\`
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOT(TRUE(), 1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: NOT(value: BOOLEAN) takes at most 1 argument, got 2
      "
    `)
  })

  it('should result in FALSE when the argument is TRUE', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=1')
    spreadsheet.set('B1', '=NOT(AS_BOOLEAN(A1))')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬───────┐
      │   │ A │ B     │
      ├───┼───┼───────┤
      │ 1 │ 1 │ FALSE │
      └───┴───┴───────┘
      "
    `)
  })

  it('should result in TRUE when the argument is falsy', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=0')
    spreadsheet.set('B1', '=NOT(AS_BOOLEAN(A1))')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬──────┐
      │   │ A │ B    │
      ├───┼───┼──────┤
      │ 1 │ 0 │ TRUE │
      └───┴───┴──────┘
      "
    `)
  })
})
