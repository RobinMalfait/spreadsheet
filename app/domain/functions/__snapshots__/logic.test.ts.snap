// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@example tests > AND > AND(TRUE(), TRUE(), FALSE()) 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > AND > AND(TRUE(), TRUE(), TRUE()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > FALSE() 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > IF > IF(FALSE(), "huge if true", "huge if false") 1`] = `
"
┌───┬─────────────────┐
│   │ A               │
├───┼─────────────────┤
│ 1 │ "huge if false" │
└───┴─────────────────┘
"
`;

exports[`@example tests > IF > IF(TRUE(), "huge if true", "huge if false") 1`] = `
"
┌───┬────────────────┐
│   │ A              │
├───┼────────────────┤
│ 1 │ "huge if true" │
└───┴────────────────┘
"
`;

exports[`@example tests > IF_ERROR > IF_ERROR(123 / 0, 0) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`@example tests > IF_ERROR > IF_ERROR(123 / 1, 0) 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ 123 │
└───┴─────┘
"
`;

exports[`@example tests > NOT > NOT(FALSE()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > NOT > NOT(TRUE()) 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > OR > OR(TRUE(), TRUE(), FALSE()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > OR > OR(TRUE(), TRUE(), TRUE()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > SWITCH > SWITCH(1, 1, "st", 2, "nd", 3, "rd", "th") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ "st" │
└───┴──────┘
"
`;

exports[`@example tests > SWITCH > SWITCH(2, 1, "st", 2, "nd", 3, "rd", "th") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ "nd" │
└───┴──────┘
"
`;

exports[`@example tests > SWITCH > SWITCH(3, 1, "st", 2, "nd", 3, "rd", "th") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ "rd" │
└───┴──────┘
"
`;

exports[`@example tests > SWITCH > SWITCH(4, 1, "st", 2, "nd", 3, "rd", "th") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ "th" │
└───┴──────┘
"
`;

exports[`@example tests > SWITCH > SWITCH(5, 1, "st", 2, "nd", 3, "rd", "th") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ "th" │
└───┴──────┘
"
`;

exports[`@example tests > TRUE() 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;
